src/extension.ts

```diff
@@ -0,0 +1,243 @@
+  "j",
+  "halt",
+  "nop",
+] as const;
+
+type OpcodeName = typeof opcodeNames[number];
+function isOpcodeName(x: string): x is OpcodeName {
+  // widen formats to string[] so indexOf(x) works
+  return (opcodeNames as readonly string[]).indexOf(x) >= 0;
+}
+
+/**
+ * Signature of an instruction
+ *
+ * containing information of input argument(s)
+ */
+type Signature = {
+  arg1?: Argument;
+  arg2?: Argument;
+};
+
+/**
+ * Class representing an input argument of an SM213 instruction
+ */
+class Argument {
+  /** name of the argument */
+  argName: string;
+  addressingModes: AddressingMode[];
+  /** description of the argument
+   *
+   * Format: <argName: addressingMode1/addressingMode2...>
+   */
+  desc: string;
+  constructor(argName: string, modes: AddressingMode[]) {
+    this.argName = argName;
+    this.addressingModes = modes;
+    // Generate desc
+    // in the format of <argName: mode1/mode2...>
+    this.desc = `<${this.argName}: `;
+    for (const mode of modes) {
+      this.desc += `${mode}/`;
+    }
+    this.desc = this.desc.replace(/.$/, ">");
+  }
+}
+
+/**
+ * Class representing an Opcode
+ */
+class Opcode {
+  name: OpcodeName;
+  signature?: Signature;
+  /**
+   * Description of the Opcode
+   *
+   * It will be displayed as the second part of the hover preview
+   */
+  description: string;
+  markdownDesc: vscode.MarkdownString;
+
+  /**
+   *
+   * @param description description of the Opcode, for which $1, $2 are used as placeholder for argument names
+   */
+  constructor(name: OpcodeName, description: string, signature?: Signature) {
+    this.name = name;
+    this.signature = signature;
+    this.description = description;
+    // Format description from signature
+    if (this.signature?.arg1) {
+      this.description = this.description.replace(
+        "$1",
+        String.raw`\<${this.signature.arg1.argName}\>`
+      );
+    }
+    if (this.signature?.arg2) {
+      this.description = this.description.replace(
+        "$2",
+        String.raw`\<${this.signature.arg2.argName}\>`
+      );
+    }
+
+    // Generate markdownDesc
+    let heading = this.name;
+    if (this.signature?.arg1) {
+      heading += ` ${this.signature?.arg1.desc}`;
+    }
+    if (this.signature?.arg2) {
+      heading += ` ${this.signature?.arg2.desc}`;
+    }
+    this.markdownDesc = new vscode.MarkdownString()
+      .appendCodeblock(heading, "asm")
+      .appendMarkdown("\n --- \n") // insert horizontal rule
+      .appendMarkdown(this.description);
+  }
+}
+
+const OPCODES = new Map<OpcodeName, Opcode>([
+  [
+    "ld",
+    new Opcode("ld", "loads value from $1 to $2", {
+      arg1: new Argument("val", ["immediate", "base+offset", "indexed"]),
+      arg2: new Argument("dest", ["register"]),
+    }),
+  ],
+  [
+    "st",
+    new Opcode("st", "stores value from register $1 to memory address $2", {
+      arg1: new Argument("source", ["register"]),
+      arg2: new Argument("dest", ["base+offset", "indexed"]),
+    }),
```

Comment:
I was real surprised about how nice the typed out ver of semantic looks like. I thought it would be quite ugly without tex but apparently this is not the case. I will try me best to add this.

Also I think machine code is definitely a QOL addition. I will definitely add it in the future.